tests/textcase/lsp/lsp_spec.lua

```diff
@@ -1,43 +1,47 @@
 local test_helpers = require("tests.test_helpers")
 local textcase = require("textcase")
 
+-- Allow Typescript language server to start.
+-- Even though the language server is attached, it doesn't mean it is ready
+-- to receive requests. Hence, we send a request and wait for response.
+-- Then we know the language server is ready.
+local wait_for_language_server_to_start = function()
+  test_helpers.execute_keys("ww") -- Move to `doSomething`
+  local hover = ""
+  test_helpers.wait_for(30 * 1000, function()
+    -- This prints one "Error detected while processing command line:" but this can be ignored
+    vim.lsp.buf_request_all(0, "textDocument/hover", vim.lsp.util.make_position_params(), function(results)
+      -- Hover will print the type definition of the variable under the cursor. Hence,
```

Comment:
Is the `type definition` the right concept to test? 

To increase speed on the future LSP test cases, we might have to setup the buffer in the `before block`. Having a single or as minimum as possible LSP setups

Then, for each test case, it will edit the existing buffer rather than setting up a buffer from scratch. There will be a small delay while the LS computes the mutated buffer and the code should await for such change. The question is, how do we detect it? I think the `type definition of the variable under the cursor` will already exist from the previous case, therefore it won't be the right indicator

Reply:
It doesn't really matter what we ask the LS I guess. And I thought the type definition was one of the easiest.

In terms of sharing a setup between tests: plenary-busted only allows for `before_each`. This means it would still run before each test. If we want to share test setup in a "before_all", we would need to migrate to proper busted for testing, which is only possible for Neovim 0.9+. See here: https://mrcjkb.dev/posts/2023-06-06-luarocks-test.html

In any case, the readiness can be tested with the type definition I'd say. All the other things need to be tested like you did it: Check for the change in a loop and wait for it to be there.

Reply:
> plenary-busted only allows for before_each

🤔 can the solution be to locate the Lsp loading and check outside the `it statement`?

> It doesn't really matter what we ask the LS I guess

Rather than LspInfo vs Type Definition, I'd like to make a point in favor of what the "[wait 1 second](https://github.com/johmsalas/text-case.nvim/blob/4ab94070fb6c5eb6aaa956f0fed8ff0d4efaa717/tests/textcase/lsp/lsp_spec.lua#L30)" statement solves. In the original implementation it is the only piece of code that is decoupled from the Lsp Load. And it is the only one that waits for the Lsp to load the doc content rather that merely the load of the Lsp itself. 

Something like:
```Lua
describe("LSP rename", function() 
    local path = "./tests/textcase/lsp/fixtures/doc.py"
    local cmd = " silent exe 'e " .. path .. "'"
    vim.cmd(cmd)
    vim.bo.filetype = "typescript"

    wait_for_language_server_to_start()

    it("Should be triggered on keybinding", function()
        local code = test_helpers.read_file("./tests/textcase/lsp/fixtures/....py")
        vim.api.nvim_buf_set_lines(0, 0, -1, true, code)

        -- The first one is not actually required with the Type Def implementation
        -- vim.wait(1000, function() end)
    end)

    it("Should ...", function()
        local code = test_helpers.read_file("./tests/textcase/lsp/fixtures/....py")
        vim.api.nvim_buf_set_lines(0, 0, -1, true, code)

        --
        vim.wait(1000, function() end)
    end)

    it("Should ...", function()
        local code = test_helpers.read_file("./tests/textcase/lsp/fixtures/....py")
        vim.api.nvim_buf_set_lines(0, 0, -1, true, code)

        --
        vim.wait(1000, function() end)
    end)
end)
```

Of course, I'd rather not to use the strongly hard-coded 1 second await

Reply:
But the wait before the assert should be enough, right? When the LSP is ready to respond with a type definition, the doc has to be loaded already.